packages/commerce-server/src/determine-purchase-type.ts

```diff
@@ -33,6 +33,98 @@ const match = (
   return Boolean(value1 && value2 && value1 === value2)
 }
 
+const {getPurchasesForUser, getPurchaseForStripeCharge} = getSdk()
+
+type PurchaseForStripeCharge = NonNullable<
+  Awaited<ReturnType<typeof getPurchaseForStripeCharge>>
+>
+type Purchase = Awaited<ReturnType<typeof getPurchasesForUser>>[number]
+type BulkCoupon = NonNullable<Pick<Purchase, 'bulkCoupon'>['bulkCoupon']>
+type BulkPurchase = Purchase & {bulkCoupon: BulkCoupon}
+type BulkPurchaseData = {
+  focalProduct: {
+    productId: string
+    focalPurchase: PurchaseForStripeCharge
+    otherTeamBulkPurchases: BulkPurchase[]
+    otherIndividualPurchases: Purchase[]
+    focalBulkPurchaseIsChronologicallyFirst: boolean
+  }
+}
+type IndividualPurchaseData = {
+  focalProduct: {
+    productId: string
+    focalPurchase: PurchaseForStripeCharge
+    otherIndividualPurchases: Purchase[]
+  }
+}
+
+const summarizePurchases = (
+  focalPurchase: PurchaseForStripeCharge,
+  allUserPurchases: Purchase[],
+):
+  | {type: 'BULK'; data: BulkPurchaseData}
+  | {type: 'INDIVIDUAL'; data: IndividualPurchaseData} => {
```

Comment:
the breakthrough for me in this implementation was separating the logic that figures out the important/relevant pieces of info about a user's purchases **FROM** the logic that determines and returns the purchase type